shared/engine/transport-shared.js

```diff
@@ -2,113 +2,90 @@
 // Classes used to handle RPCs. Ability to inject delays into calls to/from server
 import rpc from 'framed-msgpack-rpc'
 import {localLog} from '../util/forward-logs'
-import {printRPC} from '../local-debug'
+import {printRPC, printRPCStats} from '../local-debug'
 import {requestIdleCallback} from '../util/idle-callback'
-
-import type {rpcLogType} from './index.platform'
-
-const {
-  transport: {RobustTransport},
-  client: {Client: RpcClient},
-} = rpc
-const KEYBASE_RPC_DELAY_RESULT: number = process.env.KEYBASE_RPC_DELAY_RESULT
-  ? parseInt(process.env.KEYBASE_RPC_DELAY_RESULT)
-  : 0
-const KEYBASE_RPC_DELAY: number = process.env.KEYBASE_RPC_DELAY ? parseInt(process.env.KEYBASE_RPC_DELAY) : 0
-
-// Wrapped to ensure its called once
-function _makeOnceOnly<F: Function>(f: F): F {
+import * as Stats from './stats'
+
+const RobustTransport = rpc.transport.RobustTransport
+const RpcClient = rpc.client.Client
+
+// We basically always log/ensure once all the calls back and forth
+function _wrap<A1, A2, A3, A4, A5, F: (A1, A2, A3, A4, A5) => void>(options: {
+  handler: F,
+  type: string,
+  method?: string,
+  extra?: Object,
+  // we only want to enfoce a single callback on some wrapped things
+  enforceOnlyOnce: boolean,
+  // if we don't have access to the method now it should be extracted from the a1 param
+  methodInCallback: boolean,
+}): F {
+  const {handler, extra, method, type, enforceOnlyOnce, methodInCallback} = options
   let once = false
-  // $FlowIssue
-  return (...args) => {
-    if (once) {
-      rpcLog('engineInternal', 'ignoring multiple result calls', {args})
+  // $ForceType
+  const wrapped: F = (a1: A1, a2: A2, a3: A3, a4: A4, a5: A5): void => {
```

Comment:
having one wrapping function is a lot simpler to understand than all these nested calls that make wrappers